### PR TITLE
Initial attempt at rich Linear Algebra example

### PR DESCRIPTION
**Pull request by tclune**
Need further discussion with Magne regarding how to nest templates.

### CODE REVIEW

File: examples/linear_algebra/Matrix.F90 [hunk 1]
```diff
@@ -0,0 +1,274 @@
...
+     type, deferred :: T
+     pure function op(x,y) result(z)
+        type(T), intent(in) :: x, y
+        type(T)) :: z
+     end function op
+   end requirement oper
+
+   requirement elemental_oper(T, op)
+     type, deferred :: T
+     pure function op(x,y) result(z)
```
Comment (everythingfunctional) [inline examples/linear_algebra/Matrix.F90:16]: I'm guessing you meant this?

```suggestion
     elemental function op(x,y) result(z)
```

File: examples/linear_algebra/Matrix.F90 [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+
+
+   template Matrix_tmpl(T,plus,times,n)
+      requires elemental_oper(T,plus)
+      requires oper(T,times)
+      privawe
+
+      public :: Matrix
+      public :: operator(+)
+      public :: operator(*)
```
Comment (everythingfunctional) [inline examples/linear_algebra/Matrix.F90:31]: You don't have an `operator(*)` in this template. Did you mean `operator(.matmul.)`?

File: examples/linear_algebra/Matrix.F90 [hunk 3]
```diff
@@ -0,0 +1,274 @@
...
+         type(T) :: elements(n,n)
+      end type
+
+      interface operator(+)
+         procedure :: plus_matrix
+      end interface operator(+)
+
+      interface operator(.matmul.)
+         procedure :: matmul_matrix
+      end interface operator(*)
```
Comment (everythingfunctional) [inline examples/linear_algebra/Matrix.F90:43]: The names don't match here. Which one did you want?

File: examples/linear_algebra/Matrix.F90 [hunk 4]
```diff
@@ -0,0 +1,274 @@
...
+
+         interface zero
+            procedure :: zero_matrix
+         end interface zero
+
+      contains
+
+         pure function zero_matrix()
+            type(Matrix) :: zero_matrix
+            zero_matrix%elements = zero()
```
Comment (everythingfunctional) [inline examples/linear_algebra/Matrix.F90:67]: I think you meant

```suggestion
            zero_matrix%elements = zero_t()
```

File: examples/linear_algebra/Matrix.F90 [hunk 5]
```diff
@@ -0,0 +1,274 @@
...
+     type, deferred :: T
+     pure function op(x,y) result(z)
+        type(T), intent(in) :: x, y
+        type(T)) :: z
+     end function op
+   end requirement elemental_oper
+
+
+
+   template Matrix_tmpl(T,plus,times,n)
```
Comment (everythingfunctional) [inline examples/linear_algebra/Matrix.F90:24]: `n` isn't declared.

File: examples/linear_algebra/Matrix.F90 [hunk 6]
```diff
@@ -0,0 +1,274 @@
...
+
+      !======================================
+      template MatrixZero_tmpl(zero_t)
+      !--------------------------------------
+         requires elemental_oper(zero_t)
+         private
+         public :: zero
+
+         interface
+            elemental function zero_t()
```
Comment (everythingfunctional) [inline examples/linear_algebra/Matrix.F90:54]: Technically, I don't think this needs to be elemental.

File: examples/linear_algebra/Matrix.F90 [hunk 7]
```diff
@@ -0,0 +1,274 @@
...
+         interface one
+            procedure :: identity_matrix
+         end interface one
+
+      contains
+
+         pure function identity_matrix()
+            type(Matrix) :: matrix_one
+
+            matix_one%elements = 0
```
Comment (everythingfunctional) [inline examples/linear_algebra/Matrix.F90:97]: I think this needs to be

```suggestion
            matix_one%elements = zero_t()
```

File: examples/linear_algebra/Matrix.F90 [hunk 8]
```diff
@@ -0,0 +1,274 @@
...
+     end template
+     !=====================================
+
+
+     !=======================================
+     template MatrixSubtraction_tmpl(minus_t)
+     !---------------------------------------
+        requires elemental_oper(T, minus_t)
+        private
+        public :: minus_t
```
Comment (everythingfunctional) [inline examples/linear_algebra/Matrix.F90:111]: I think you meant

```suggestion
        public :: minus
```

File: examples/linear_algebra/Matrix.F90 (hunk 1 above)
Comment (tclune) [reply examples/linear_algebra/Matrix.F90]: Yes - originally just had one requirement and then decided to split and updated the wrong one.

File: examples/linear_algebra/Matrix.F90 (hunk 3 above)
Comment (tclune) [reply examples/linear_algebra/Matrix.F90:43]: Not sure what you mean.   These do not need to match, but my intent was `OPERATOR(*)` in any event.

**Pull request merged**